tests/providers/openstack/os_virtual_machine_test.py

```diff
@@ -0,0 +1,103 @@
+
+  def _CreateBenchmarkSpecFromYaml(self, yaml_string,
+                                   benchmark_name=_BENCHMARK_NAME):
+    config = configs.LoadConfig(yaml_string, {}, benchmark_name)
+    return self._CreateBenchmarkSpecFromConfigDict(config, benchmark_name)
+
+  def _CreateBenchmarkSpecFromConfigDict(self, config_dict, benchmark_name):
+    config_spec = benchmark_config_spec.BenchmarkConfigSpec(
+        benchmark_name,
+        flag_values=FLAGS,
+        **config_dict)
+    benchmark_module = next((b for b in linux_benchmarks.BENCHMARKS
+                             if b.BENCHMARK_NAME == benchmark_name))
+    return benchmark_spec.BenchmarkSpec(benchmark_module, config_spec, _URI)
+
+
+class OpenStackVirtualMachineTest(BaseOpenStackNetworkTest):
+
+
+  @mock.patch.object(os_virtual_machine.OpenStackVirtualMachine, '_CheckNetworkExists', return_value=_network_true)
+  def test_CheckFloatingIPNetworkExistsWithTrue(self, mock_CheckNetworkExists, flags=FLAGS):
+    mock_CheckNetworkExists.return_value = _network_true
+
+    flags.ignore_package_requirements = True
+    spec = self._CreateBenchmarkSpecFromYaml(_CFG_DEFAULT_DEFAULT)
+    spec.disable_interrupt_moderation = False
+    spec.disable_rss = False
+    spec.zone = "test-zone"
+    spec.cidr = "192.164.1.0/24"
+    spec.machine_type = "Test_machine_type"
+    spec.gpu_count = "1"
+    spec.gpu_type = "test-gpu-type"
+    spec.image = "test-image"
+    spec.install_packages = "None"
+    spec.background_cpu_threads = "None"
+    spec.background_network_mbits_per_sec = "1"
+    spec.background_network_ip_type = "None"
+    spec.vm_metadata = {}
+    openstackVM = os_virtual_machine.OpenStackVirtualMachine(spec)
+    network = openstackVM._CheckFloatingIPNetworkExists('External')
+    self.assertEqual(_network_true, network)
+
+
+  @mock.patch.object(os_virtual_machine.OpenStackVirtualMachine, '_CheckNetworkExists', return_value=_network_external)
+  def test_CheckFloatingIPNetworkExistsWithExternal(self, mock_CheckNetworkExists, flags=FLAGS):
+    mock_CheckNetworkExists.return_value = _network_external
+
+    flags.ignore_package_requirements = True
+    spec = self._CreateBenchmarkSpecFromYaml(_CFG_DEFAULT_DEFAULT)
+    spec.disable_interrupt_moderation = False
+    spec.disable_rss = False
+    spec.zone = "test-zone"
+    spec.cidr = "192.164.1.0/24"
+    spec.machine_type = "Test_machine_type"
+    spec.gpu_count = "1"
+    spec.gpu_type = "test-gpu-type"
+    spec.image = "test-image"
+    spec.install_packages = "None"
+    spec.background_cpu_threads = "None"
+    spec.background_network_mbits_per_sec = "1"
+    spec.background_network_ip_type = "None"
+    spec.vm_metadata = {}
+    openstackVM = os_virtual_machine.OpenStackVirtualMachine(spec)
+    network = openstackVM._CheckFloatingIPNetworkExists('External')
+    self.assertEqual(_network_external, network)
+
+
+if __name__ == '__main__':
+  unittest.main()
```

Comment:
Add a negative test, where you expect the test to fail, such as {'router:external': 'Foobar'}

Reply:
Done :)